tooling/profiler/src/flamegraph.rs

```diff
@@ -88,37 +89,49 @@ fn generate_folded_sorted_lines<'files, F: AcirField>(
 
-    samples.into_iter().for_each(|sample| {
-        let mut location_names: Vec<String> = sample
-            .call_stack
-            .into_iter()
-            .flat_map(|opcode_location| {
-                debug_symbols.opcode_location(&opcode_location).unwrap_or_else(|| {
-                    if let (Some(brillig_function_id), Some(brillig_location)) =
-                        (sample.brillig_function_id, opcode_location.to_brillig_location())
-                    {
-                        let brillig_locations =
-                            debug_symbols.brillig_locations.get(&brillig_function_id);
-                        if let Some(brillig_locations) = brillig_locations {
-                            brillig_locations.get(&brillig_location).cloned().unwrap_or_default()
-                        } else {
-                            vec![]
-                        }
-                    } else {
-                        vec![]
-                    }
-                })
-            })
-            .map(|location| location_to_callsite_label(location, files))
-            .collect();
-
-        if location_names.is_empty() {
-            location_names.push("unknown".to_string());
+    let mut resolution_cache: HashMap<OpcodeLocation, Vec<String>> = HashMap::default();
+    for sample in samples {
+        let mut location_names = Vec::with_capacity(sample.call_stack.len());
+        for opcode_location in sample.call_stack {
+            let callsite_labels =
+                if let Some(callsite_labels) = resolution_cache.get(&opcode_location) {
+                    callsite_labels.clone()
+                } else {
+                    let source_locations =
+                        debug_symbols.opcode_location(&opcode_location).unwrap_or_else(|| {
+                            if let (Some(brillig_function_id), Some(brillig_location)) =
+                                (sample.brillig_function_id, opcode_location.to_brillig_location())
+                            {
+                                let brillig_locations =
+                                    debug_symbols.brillig_locations.get(&brillig_function_id);
+                                if let Some(brillig_locations) = brillig_locations {
+                                    brillig_locations
+                                        .get(&brillig_location)
+                                        .cloned()
+                                        .unwrap_or_default()
+                                } else {
+                                    vec![]
+                                }
+                            } else {
+                                vec![]
+                            }
+                        });
+                    let callsite_labels: Vec<_> = source_locations
+                        .into_iter()
+                        .map(|location| location_to_callsite_label(location, files))
+                        .collect();
```

Comment:
Perhaps now that there is caching, it would be helpful to put this into a function that turns an opcode location into callsite labels using the debug symbols. Then what's left here could be as simple as 
```rust
let callsite_labels = resolution_cache.entry(opcode_location).or_insert_with(|| 
  find_callsites(&debug_symbols, sample.brillig_function_id.as_ref(), &opcode_location
).clone();`